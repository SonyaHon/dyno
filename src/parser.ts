import { Lexer } from "./lexer";
import { Token, TokenType } from "./token";

export interface Statement {}

export interface Expression {}

export class Identifier {
  constructor(public readonly token: Token, public readonly name: string) {}
}

export class NumberLiteral implements Expression {
  constructor(public readonly token: Token, public readonly value: number) {}
}

export class Type {}

export class DefineConstant implements Statement {
  constructor(
    public token: Token,
    public name: Identifier,
    public value: Expression
  ) {}
}

export class Module {
  public statements: Statement[] = [];
}

export class Parser {
  private currentToken: Token;
  private peekToken: Token;

  static ParseText(input: string): Module {
    const parser = new Parser(new Lexer(input));
    return parser.parse();
  }

  constructor(public lexer: Lexer) {
    this.currentToken = this.lexer.nextToken();
    this.peekToken = this.lexer.nextToken();
  }

  private iterate() {
    this.currentToken = this.peekToken;
    this.peekToken = this.lexer.nextToken();
  }

  private parseIdentStarting() {
    if (this.peekToken.type === TokenType.ColonColon) {
      const token = this.currentToken;
      const ident = new Identifier(
        this.currentToken,
        this.currentToken.literal
      );

      this.iterate();
      //@ts-ignore
      if (this.peekToken.type === TokenType.Number) {
        const numberToken = this.peekToken;
        this.iterate();
        this.iterate();
        return new DefineConstant(
          token,
          ident,
          new NumberLiteral(numberToken, Number(numberToken.literal))
        );
      }
    }
    throw new Error("Unreachable");
  }

  parse(): Module {
    const module = new Module();

    while (this.currentToken.type !== TokenType.Eof) {
      let statement = null;

      if (this.currentToken.type === TokenType.Ident) {
        statement = this.parseIdentStarting();
      }

      if (statement) {
        module.statements.push(statement);
      }
    }

    return module;
  }
}
