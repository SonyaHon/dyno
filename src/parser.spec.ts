import { describe, expect, test } from "bun:test";
import { Parser } from "./parser";

describe("parser", function () {
  test("const number literal definition", function () {
    const input = "x :: 69";
    const module = Parser.ParseText(input);
    expect(module).toMatchSnapshot();
  });
});
