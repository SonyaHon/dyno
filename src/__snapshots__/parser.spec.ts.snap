// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parser const number literal definition 1`] = `
Module {
  "statements": [
    DefineConstant {
      "name": Identifier {
        "name": "x",
        "token": Token {
          "literal": "x",
          "type": "Ident",
        },
      },
      "token": Token {
        "literal": "x",
        "type": "Ident",
      },
      "value": NumberLiteral {
        "token": Token {
          "literal": "69",
          "type": "Number",
        },
        "value": 69,
      },
    },
  ],
}
`;
